### ISSUE TITLE
Redenoising Sampling is confusing in paper, can you clarify?

### ISSUE THREAD

**Issue by Cortexelus** [xie-lab-ml/Golden-Noise-for-Diffusion-Models#5]
Amazing [paper](https://arxiv.org/pdf/2411.09502) you guys. 

The definition of redenoising sampling seems ambiguous to me (& it was also confusing for R1 and ChatGPT) so I'd like some help with clarification.

When you write
`x't=DDIM-Inversion(DDIM(xt))`

_Does DDIM mean a single step, or the full generation schedule?_

Does this mean, for SDXL:
- You start by running the typical generation process to create an image. 
- From gaussian noise you run DDIM for 10 steps (with CFG 10) like you typically would. 
- Then, you run DDIM-Inversion for 10 steps (CFG 1, aka without CFG) to obtain the new noise. 
- You only run this cycle once. 

Or, does this mean you stay near the noisy end of the schedule:
- You start with gaussian noise. You run DDIM with for 1 step (with CFG 10)
- Then 1 step of DDIM-Inversion (without CFG) to get a noise
- And you run this cycle 10x times, to find a noise very much stuck to its fixed point

And in **A.4. Hyper-parameter Settings**

"To collect training data, the inference steps are
- 10 for SDXL with DDIM inverse scheduler
- 4 for Dreamshaper-xl-v2-turbo with DPMSolver inverse scheduler
- 10 for Hunyuan-DiT with DDIM inverse scheduler"
...
... all these noise pairs are collected with inference step 10
to construct NPDs...."

I'm confused if Dreamshaper-xl-v2-turbo used 4 or 10 steps during redenoising. Are these two different concepts of step count? 

Thanks

**Comment by Klayand** [comment]
Hello Cortexelus,

Thanks for pointing out.

In the whole paper, re-denosing sampling means only the first step:
x’_T=DDIM-Inversion(DDIM(x_T)) = DDIM-Inversion(x_{T-1})

For in A.4. Hyper-parameter Settings:

The reason why we collect the noise prompt datasets using different steps is for efficiency. That means you can collect the noise pairs with different inference steps, also with different deterministic samplers. 

10 for SDXL, 4 for Dreamshaper-xl-v2-turbo, and 10 for Hunyuan-DiT to help us quickly collect the noise pairs.

If you have any questions, please let me know.

**Comment by Klayand** [comment]
By the way, if you want to learn more about the re-denosing sampling technique, I recommend one paper of our group accepted in ICLR 2025.

Zigzag Diffusion Sampling: The Path to Success ls Zigzag
https://openreview.net/forum?id=MKvQH1ekeY&referrer=%5BAuthor%20Console%5D(%2Fgroup%3Fid%3DICLR.cc%2F2025%2FConference%2FAuthors%23your-submissions)

Re-denoising sampling is a special case of it. 

The inference code of Zigzag is also public:
https://github.com/xie-lab-ml/Zigzag-Diffusion-Sampling

**Comment by Cortexelus** [comment]
Ah thank you. The Zigzag paper is awesome too!

**Comment by Cortexelus** [comment]
If you fully invert a Zigzag sampled image back to noise, is this a
superior Golden Noise for the prompt?

Is there any gain in repeating this zigzag-invert cycle?

And thirdly, we found superior success using the APG alternative to CFG
https://arxiv.org/abs/2410.02416
This uses only the perpendicular component for CFG. It reduces saturation.
Also makes high CFG values more stable. I wonder how APG might affect
inversion, Zigzag, and golden noise.

**Comment by FromA2Z** [comment]
> If you fully invert a Zigzag sampled image back to noise, is this a
> superior Golden Noise for the prompt?
> 
> Is there any gain in repeating this zigzag-invert cycle?
> 
> And thirdly, we found superior success using the APG alternative to CFG
> https://arxiv.org/abs/2410.02416
> This uses only the perpendicular component for CFG. It reduces saturation.
> Also makes high CFG values more stable. I wonder how APG might affect
> inversion, Zigzag, and golden noise.
> […](#)

Thanks for your question—that's awesome! Both Z-Sampling and Golden Noise are integral parts of our series of work. Z-Sampling focuses on optimizing the sampling path, while Golden Noise enhances the initial noise through a lightweight network.

Q:Is there any gain in repeating this zigzag-invert cycle?

A:Indeed, fully inverting the process can significantly degrade the results, as demonstrated in Appendix E.2 of Z-Sampling through our experiments (see Figure 25, Table 15, and Table 16 for details). This is because end-to-end methods tend to amplify inversion errors, leading to negative gains (as explained in Theorem 1 and Theorem 2). Therefore, a step-by-step truncation approach is a better choice.

Q:Is there any gain in repeating this zigzag-invert cycle?

Yes, regarding multiple iterations, this is a hyperparameter. In the official code of Z-Sampling, we have provided an interface for this parameter (i.e., T_max). The default setting is T_max=1, and our experiments show that one or two iterations yield the best results. When the number of iterations increases, the generated results tend to degrade or break down.

Q:Regarding the APG.

Undoubtedly, APG is an awesome piece of work! Especially its idea of decoupling the predicted scores into horizontal and vertical components—it's truly brilliant.

Indeed, while working on Z-Sampling, we tried integrating APG but faced an issue: the norms of the horizontal and vertical components differed too much, making the optimized scores after decoupling nearly identical to the original ones, resulting in minimal improvement.

We speculate that this might be because we only used the "decoupling" component in APG and did not incorporate other elements such as adaptive scaling.

We hope these discussions spark some new ideas—have a great day!

**Comment by Cortexelus** [comment]
Great answers thank you. 
We adaptive scaling implemented w/ APG so I will try this after implementing Z-sampling.
